### PR TITLE
bpo-28416: Break reference cycles in Pickler and Unpickler subclasses

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
with the persistent_id() and persistent_load() methods.

https://bugs.python.org/issue28416

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
@vstinner, @pitrou, could you please look at this?

File: Lib/test/test_pickle.py [hunk 1]
```diff
@@ -112,6 +113,66 @@ class PyIdPersPicklerTests(AbstractIdentityPersistentPicklerTests,
     pickler = pickle._Pickler
     unpickler = pickle._Unpickler
 
+    @support.cpython_only
+    def test_pickler_reference_cycle(self):
+        def check(Pickler):
+            for bin in 0, 1:
```
Comment (pitrou) [inline Lib/test/test_pickle.py]: Hmm, what is `bin`? Do you mean `protocol`?

Comment (serhiy-storchaka) [reply Lib/test/test_pickle.py]: A flag text/binary. Later it became a protocol, but the attribute `bin` still is used in sources (it is equal `protocol != 0`). Here it is enough to test only one binary protocol.

Comment (pitrou) [reply Lib/test/test_pickle.py]: Yes... It's still a bit cryptic though. And it doesn't cost anything to loop over all protocols, right? :-)

Comment (serhiy-storchaka) [reply Lib/test/test_pickle.py]: Done.

File: Modules/_pickle.c [hunk 2]
```diff
@@ -360,6 +360,81 @@ _Pickle_FastCall(PyObject *func, PyObject *obj)
...
+
+    /* *method and *unbound should be consistent.  All refcount decrements
+       should be occurred after setting *unbound and *method. */
+    *unbound = 0;
+    Py_CLEAR(*method);
+
+    func = _PyObject_GetAttrId(self, name);
+    if (func == NULL) {
+        *unbound = 0;
+        Py_CLEAR(*method);
```
Comment (pitrou) [inline Modules/_pickle.c]: The two lines above seem redundant with the function beginning.

File: Modules/_pickle.c [hunk 3]
```diff
@@ -360,6 +360,81 @@ _Pickle_FastCall(PyObject *func, PyObject *obj)
 
 /*************************************************************************/
 
+static int
+init_method_ref(PyObject *self, _Py_Identifier *name,
```
Comment (pitrou) [inline Modules/_pickle.c]: Add a comment explaining that the purpose is to avoid reference cycles through `self`?

File: Modules/_pickle.c [hunk 4]
```diff
@@ -360,6 +360,81 @@ _Pickle_FastCall(PyObject *func, PyObject *obj)
...
+
+static PyObject *
+bound_method(PyObject *self, PyObject *func, int unbound)
+{
+    if (unbound) {
+        descrgetfunc f = func->ob_type->tp_descr_get;
+        if (f != NULL) {
+            return f(func, self, (PyObject *)Py_TYPE(self));
+        }
+    }
```
Comment (pitrou) [inline Modules/_pickle.c]: Why does this fall through? May we return a non-bound method when `unbound` is true?

File: Modules/_pickle.c [hunk 5]
```diff
@@ -360,6 +360,81 @@ _Pickle_FastCall(PyObject *func, PyObject *obj)
 
 /*************************************************************************/
 
+static int
```
Comment (pitrou) [inline Modules/_pickle.c]: I wonder if the role of those three functions would be clearer if they took a `{PyObject *func, int unbound}` structure.

Comment (serhiy-storchaka) [reply Modules/_pickle.c]: I don't know. This function returns a pair: a pointer to bound or unbound method, and a flag for distinguishing them. There are other functions that return the same pair: _PyObject_GetMethod() and lookup_maybe_method() in typeobject.c. But the interface is different. _PyObject_GetMethod() returns a flag and sets a pointer by reference, lookup_maybe_method() returns a pointer and sets a flag by reference, this functions sets both pointer and flag by reference (and it also decrement the refcount of the old pointer). I tried different variants, but I don't know what interface is better. Well, will try to use a structure.

File: Modules/_pickle.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Modules/_pickle.c]: Actually lines at function beginning are redundant.

File: Modules/_pickle.c (hunk 4 above)
Comment (serhiy-storchaka) [reply Modules/_pickle.c]: Hmm, interesting question.

**Comment by serhiy-storchaka** [comment]
Using a structure will lead to larger diff. Currently the code that just examine `pers_func` and handle it as a regular object reference (deallocators, etc) is not changed.

But I tried other refactoring. Now an additional parameter is not a flag, but a borrowed reference to self or NULL. This simplifies some calls.

File: Modules/_pickle.c [hunk 6]
```diff
@@ -360,6 +360,69 @@ _Pickle_FastCall(PyObject *func, PyObject *obj)
...
+        return 0;
+    }
+    else {
+        *method_self = NULL;
+        Py_XSETREF(*method_func, func);
+        return 0;
+    }
+}
+
+/* Bound a method if it was deconstructed */
```
Comment (pitrou) [inline Modules/_pickle.c]: Infinitive is "bind", not "bound".

**Comment by pitrou** [comment]
Thank you. This looks good to me now.

**Comment by serhiy-storchaka** [comment]
Thanks Antoine!

**Pull request merged**